### DISCUSSION TITLE
DriverException when using pessimistic LockMode in combination with populate

### DISCUSSION THREAD

**Discussion by decline** [mikro-orm/mikro-orm#5404] [General]
### Describe the bug

When using `@mikro-orm/postgresql` as database driver, there seems to be an issue when using `lockMode: LockMode.PESSIMISTIC_WRITE` in combination with `populate: ['xyz']` option.

In the [reproduction repository](https://github.com/decline/mikro-orm-lock-mode-populate-issue), you can see the affected line in the `example.test.ts` on line 58:

```ts
const user = await orm.em.findOneOrFail(User, { email: 'foo' }, { lockMode: LockMode.PESSIMISTIC_WRITE, populate: ['subscription'] });
```

This will throw the following exception:

```sh
DriverException: select "u0".*, "s1"."name" as "s1__name", "s1"."credits" as "s1__credits" from "user" as "u0" left join "subscription" as "s1" on "u0"."subscription_name" = "s1"."name" where "u0"."email" = 'foo' limit 1 for update - FOR UPDATE cannot be applied to the nullable side of an outer join
```

**Hint:** This was previously working in `v5.7.13`. And it will still work once I change the `loadStrategy` back to `select-in`, so I guess it has something to do with this change?

### Reproduction

https://github.com/decline/mikro-orm-lock-mode-populate-issue

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.1.13-dev.15

### Node.js version

20.10.0

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
Sounds like the same as #5125, try using `lockTableAliases` option.

**Reply by decline** [reply]
Thanks for the info, but when adding `lockTableAliases: ['user']` to the options, I'm getting the following error:

```sh
TableNotFoundException: select "u0".*, "s1"."name" as "s1__name", "s1"."credits" as "s1__credits" from "user" as "u0" left join "subscription" as "s1" on "u0"."subscription_name" = "s1"."name" where "u0"."email" = 'foo' limit 1 for update of "user" - relation "user" in FOR UPDATE clause not found in FROM clause
```

It seems it's not using the alias in the FROM clause. Using `lockTableAliases: ['u0']` would work, but that does not feel safe.

**Reply by B4nan** [reply]
The option is for aliases (I hope that should be clear from the option name), not table names - as you can have multiple joins of the same table and you need to distinguish those.

**Reply by B4nan** [reply]
I don't think passing `user` would be any better. For QB you could use `qb.alias` to get the root entity alias, for `em.find` there is no way to have this dynamic. I am open for suggestions on how to make this better, maybe we could have another option where you would be able to pass join paths instead - that way it is relatively safe to infer the alias out of it (although technically there can be still multiple join branches for one path).

Something like this:

```ts
lockPaths: [
  'User', // root entity
  'User.books.author', //  nested relation, which would also resolve to User entity, but with different alias based on the path
],
```

**Reply by decline** [reply]
I understand - I guess in my scenario using something like `const userAlias = orm.em.createQueryBuilder(User).alias;` and then using it as `lockTableAliases: [userAlias]` should be _relatively_ safe.

But I agree, it will get tricky once you have multiple join paths or maybe 2 entities that start with the same letter, like `User` and `UserSubscription`, for which the aliases could be equally `u0` or `u1`, depending on which side of the relation they will be used.

**Reply by B4nan** [reply]
Root entity always has the `0` suffix, so its not that bad, the real problem comes with distinguishing relations.

**Comment by B4nan** [comment]
Should work natively in next version: https://github.com/mikro-orm/mikro-orm/commit/3fdb077893144111cf4d8b2ea5a22271d0f405df
